File: rust/mp5-codec/src/mp5c/fixtures.rs
//! Synthetic PCM fixtures for MP5-C benchmarks (no external files required).

pub struct Fixture {
    pub name: &'static str,
    pub samples: Vec<i16>,
    pub channels: u8,
    pub sample_rate: u32,
}

pub fn all_fixtures() -> Vec<Fixture> {
    vec![
        sine_stereo_2s(),
        silence_stereo_2s(),
        white_noise_stereo_2s(),
        pink_noise_stereo_2s(),
        drums_stereo_5s(),
        vocal_like_stereo_5s(),
        music_like_stereo_30s(),
        sine_stereo_0_5s(),
    ]
}

fn sr() -> usize {
    44100
}

fn sine_stereo_2s() -> Fixture {
    let n = sr() * 2;
    let mut s = vec![0i16; n * 2];
    for i in 0..n {
        s[i * 2] = ((440.0 * 2.0 * std::f32::consts::PI * i as f32 / sr() as f32).sin() * 14000.0) as i16;
        s[i * 2 + 1] = ((330.0 * 2.0 * std::f32::consts::PI * i as f32 / sr() as f32).cos() * 14000.0) as i16;
    }
    Fixture {
        name: "sine_2s",
        samples: s,
        channels: 2,
        sample_rate: sr() as u32,
    }
}

fn sine_stereo_0_5s() -> Fixture {
    let n = sr() / 2;
    let mut s = vec![0i16; n * 2];
    for i in 0..n {
        s[i * 2] = ((440.0 * 2.0 * std::f32::consts::PI * i as f32 / sr() as f32).sin() * 14000.0) as i16;
        s[i * 2 + 1] = s[i * 2];
    }
    Fixture {
        name: "sine_0.5s_short",
        samples: s,
        channels: 2,
        sample_rate: sr() as u32,
    }
}

fn silence_stereo_2s() -> Fixture {
    Fixture {
        name: "silence_2s",
        samples: vec![0i16; sr() * 2 * 2],
        channels: 2,
        sample_rate: sr() as u32,
    }
}

fn white_noise_stereo_2s() -> Fixture {
    let n = sr() * 2;
    let mut s = vec![0i16; n * 2];
    let mut state = 0x12345678u32;
    for i in 0..n * 2 {
        state = state.wrapping_mul(1664525).wrapping_add(1013904223);
        s[i] = ((state >> 16) as i32 % 32767) as i16;
    }
    Fixture {
        name: "white_noise_2s",
        samples: s,
        channels: 2,
        sample_rate: sr() as u32,
    }
}

fn pink_noise_stereo_2s() -> Fixture {
    let n = sr() * 2;
    let mut s = vec![0i16; n * 2];
    let mut b0 = 0.0f32;
    let mut b1 = 0.0f32;
    let mut b2 = 0.0f32;
    let mut state = 1u32;
    for i in 0..n * 2 {
        state = state.wrapping_mul(1664525).wrapping_add(1013904223);
        let w = (state as f32 / u32::MAX as f32) * 2.0 - 1.0;
        b0 = 0.99765 * b0 + w * 0.0990460;
        b1 = 0.96300 * b1 + w * 0.2965164;
        b2 = 0.57000 * b2 + w * 1.0526913;
        let pink = b0 + b1 + b2 + w * 0.1848;
        s[i] = (pink.clamp(-1.0, 1.0) * 12000.0) as i16;
    }
    Fixture {
        name: "pink_noise_2s",
        samples: s,
        channels: 2,
        sample_rate: sr() as u32,
    }
}

fn drums_stereo_5s() -> Fixture {
    let n = sr() * 5;
    let mut s = vec![0i16; n * 2];
    for i in 0..n {
        let t = i as f32 / sr() as f32;
        let beat = (t * 2.0).floor();
        let env = (1.0 - (t * 2.0 - beat)).max(0.0);
        let kick = (env.powf(3.0) * (t * 80.0).sin() * 28000.0) as i16;
        let snare = if (beat as i32 % 2) == 1 {
            (((t * 200.0).sin() * 0.3) * 18000.0) as i16
        } else {
            0
        };
        let v = kick.saturating_add(snare);
        s[i * 2] = v;
        s[i * 2 + 1] = v;
    }
    Fixture {
        name: "drums_5s",
        samples: s,
        channels: 2,
        sample_rate: sr() as u32,
    }
}

fn vocal_like_stereo_5s() -> Fixture {
    let n = sr() * 5;
    let mut s = vec![0i16; n * 2];
    for i in 0..n {
        let t = i as f32 / sr() as f32;
        let f0 = 220.0 + 30.0 * (t * 3.0).sin();
        let mut v = 0.0f32;
        for h in 1..=6 {
            v += (f0 * h as f32 * t * std::f32::consts::TAU).sin() / h as f32;
        }
        let formant = (t * 2500.0).sin() * 0.15;
        let amp = 0.5 + 0.5 * (t * 2.5).sin().max(0.0);
        let sample = ((v + formant) * amp * 10000.0) as i16;
        s[i * 2] = sample;
        s[i * 2 + 1] = (sample as f32 * 0.92) as i16;
    }
    Fixture {
        name: "vocal_like_5s",
        samples: s,
        channels: 2,
        sample_rate: sr() as u32,
    }
}

fn music_like_stereo_30s() -> Fixture {
    let n = sr() * 30;
    let mut s = vec![0i16; n * 2];
    for i in 0..n {
        let t = i as f32 / sr() as f32;
        let kick = (t * 2.0 * std::f32::consts::PI).sin() * 12000.0;
        let bass = (t * 110.0 * std::f32::consts::TAU).sin() * 18000.0;
        let lead = (t * 440.0 * std::f32::consts::TAU).sin() * 6000.0;
        let pad = (t * 220.0 * std::f32::consts::TAU).sin() * 4000.0;
        let v = (kick + bass + lead + pad) as i16;
        s[i * 2] = v;
        s[i * 2 + 1] = (v as f32 * 0.97) as i16;
    }
    Fixture {
        name: "music_like_30s",
        samples: s,
        channels: 2,
        sample_rate: sr() as u32,
    }
}

File: tools/audio-lab/fixtures.mjs
// MP5 Audio Quality Lab — synthetic fixture generator.
// All fixtures are generated deterministically; nothing is read from disk and
// no copyrighted audio is ever used. Each fixture is stereo i16 at 44.1 kHz
// unless noted. Categories mirror docs/MP5_AUDIO_QUALITY_LAB.md.

const SR = 44100;
const TAU = Math.PI * 2;

// Deterministic LCG so noise fixtures are reproducible across runs/machines.
function lcg(seed) {
  let state = seed >>> 0;
  return () => {
    state = (Math.imul(state, 1664525) + 1013904223) >>> 0;
    return state / 0xffffffff; // [0,1)
  };
}

function clampI16(x) {
  if (x > 32767) return 32767;
  if (x < -32768) return -32768;
  return x | 0;
}

function stereo(frames, fn) {
  const s = new Int16Array(frames * 2);
  for (let i = 0; i < frames; i++) {
    const [l, r] = fn(i);
    s[i * 2] = clampI16(Math.round(l));
    s[i * 2 + 1] = clampI16(Math.round(r));
  }
  return s;
}

function fixture(name, category, samples, note = "") {
  return { name, category, samples, channels: 2, sampleRate: SR, note };
}

// ── individual generators ───────────────────────────────────────────────────

function silence() {
  return fixture("silence", "silence", new Int16Array(SR * 2 * 2), "2s digital silence");
}

function quietSine() {
  // ~-40 dBFS — the canonical hiss exposure case.
  const amp = 0.01 * 32767;
  return fixture(
    "quiet_sine",
    "tonal",
    stereo(SR * 2, (i) => {
      const v = Math.sin((TAU * 440 * i) / SR) * amp;
      return [v, v];
    }),
    "2s 440Hz at -40 dBFS",
  );
}

function loudSine() {
  const amp = 0.95 * 32767;
  return fixture(
    "loud_sine",
    "tonal",
    stereo(SR * 2, (i) => [
      Math.sin((TAU * 440 * i) / SR) * amp,
      Math.cos((TAU * 330 * i) / SR) * amp,
    ]),
    "2s near-full-scale tones",
  );
}

function sweptSine() {
  // Log sweep 50Hz → 18kHz over 3s.
  const f0 = 50, f1 = 18000, dur = 3;
  const k = Math.log(f1 / f0) / dur;
  const amp = 0.6 * 32767;
  return fixture(
    "swept_sine",
    "tonal",
    stereo(SR * dur, (i) => {
      const t = i / SR;
      const phase = (TAU * f0 * (Math.exp(k * t) - 1)) / k;
      const v = Math.sin(phase) * amp;
      return [v, v];
    }),
    "50Hz→18kHz log sweep",
  );
}

function pinkNoise() {
  const rng = lcg(0x9e3779b1);
  let b0 = 0, b1 = 0, b2 = 0;
  return fixture(
    "pink_noise",
    "noise",
    stereo(SR * 2, () => {
      const w = rng() * 2 - 1;
      b0 = 0.99765 * b0 + w * 0.099046;
      b1 = 0.963 * b1 + w * 0.2965164;
      b2 = 0.57 * b2 + w * 1.0526913;
      const p = (b0 + b1 + b2 + w * 0.1848) * 0.35 * 32767;
      return [p, p * 0.98];
    }),
    "2s pink noise",
  );
}

function whiteNoise() {
  const rng = lcg(0x12345678);
  const amp = 0.4 * 32767;
  return fixture(
    "white_noise",
    "noise",
    stereo(SR * 2, () => [(rng() * 2 - 1) * amp, (rng() * 2 - 1) * amp]),
    "2s white noise (decorrelated)",
  );
}

function impulse() {
  const s = new Int16Array(SR * 1 * 2);
  s[SR * 2] = 32767; // single full-scale impulse at t=0.5s, otherwise silence
  s[SR * 2 + 1] = 32767;
  return fixture("impulse", "transient", s, "single full-scale impulse in silence");
}

function kickSnare() {
  return fixture(
    "kick_snare",
    "transient",
    stereo(SR * 4, (i) => {
      const t = i / SR;
      const beat = Math.floor(t * 2);
      const env = Math.max(0, 1 - (t * 2 - beat));
      const kick = Math.pow(env, 3) * Math.sin(t * 80) * 28000;
      const snare = beat % 2 === 1 ? Math.sin(t * 1800) * env * 9000 : 0;
      const v = kick + snare;
      return [v, v];
    }),
    "4s kick/snare-like transients",
  );
}

function bassLoop() {
  return fixture(
    "bass_loop",
    "musical",
    stereo(SR * 4, (i) => {
      const t = i / SR;
      const note = [55, 55, 73.42, 65.41][Math.floor(t) % 4];
      const env = 0.5 + 0.5 * Math.sin(TAU * 2 * t);
      const v = (Math.sin(TAU * note * t) + 0.3 * Math.sin(TAU * note * 2 * t)) * env * 16000;
      return [v, v];
    }),
    "4s bass-heavy loop",
  );
}

function vocalLike() {
  return fixture(
    "vocal_like",
    "musical",
    stereo(SR * 4, (i) => {
      const t = i / SR;
      const f0 = 220 + 30 * Math.sin(t * 3);
      let v = 0;
      for (let h = 1; h <= 6; h++) v += Math.sin(TAU * f0 * h * t) / h;
      const formant = Math.sin(t * 2500) * 0.15;
      const amp = 0.4 + 0.4 * Math.max(0, Math.sin(t * 2.5));
      const s = (v + formant) * amp * 9000;
      return [s, s * 0.92];
    }),
    "4s vocal-like harmonic signal",
  );
}

function stereoWidth() {
  const rng = lcg(0xcafebabe);
  return fixture(
    "stereo_width",
    "spatial",
    stereo(SR * 2, (i) => {
      const t = i / SR;
      const carrier = Math.sin(TAU * 300 * t) * 12000;
      const wob = Math.sin(TAU * 3 * t) * 8000;
      const noise = (rng() * 2 - 1) * 2000;
      return [carrier + wob + noise, carrier - wob - noise];
    }),
    "2s wide/decorrelated stereo",
  );
}

function reverbTail() {
  // A burst that decays into near-silence — the canonical reverb-tail hiss test.
  const rng = lcg(0x5eed1234);
  return fixture(
    "reverb_tail",
    "decay",
    stereo(SR * 3, (i) => {
      const t = i / SR;
      let body = 0;
      for (let h = 1; h <= 5; h++) body += Math.sin(TAU * 220 * h * t) / h;
      const tail = Math.exp(-t * 3.0); // long exponential decay toward silence
      const diffuse = (rng() * 2 - 1) * 0.15;
      const v = (body + diffuse) * tail * 14000;
      return [v, v * 0.96];
    }),
    "3s reverb-tail decay to near-silence",
  );
}

function denseMusic() {
  const rng = lcg(0xabcd1234);
  return fixture(
    "dense_music",
    "musical",
    stereo(SR * 6, (i) => {
      const t = i / SR;
      const kickEnv = Math.pow(Math.max(0, 1 - (t * 2 - Math.floor(t * 2))), 3);
      const kick = kickEnv * Math.sin(t * 70) * 14000;
      const bass = Math.sin(TAU * 110 * t) * 10000;
      const lead = Math.sin(TAU * 440 * t) * 5000;
      const pad = Math.sin(TAU * 277 * t) * 3500;
      const hat = (rng() * 2 - 1) * 1500 * (Math.floor(t * 8) % 2);
      const l = kick + bass + lead + pad + hat;
      const r = kick + bass + lead * 0.9 + pad * 1.1 - hat;
      return [l, r];
    }),
    "6s dense music-like loop",
  );
}

/** Full synthetic fixture set used by the lab. */
export function allFixtures() {
  return [
    silence(),
    quietSine(),
    loudSine(),
    sweptSine(),
    pinkNoise(),
    whiteNoise(),
    impulse(),
    kickSnare(),
    bassLoop(),
    vocalLike(),
    stereoWidth(),
    reverbTail(),
    denseMusic(),
  ];
}

/** Curated, short subset for listening exports (keeps WAV output small). */
export function listeningFixtures() {
  return allFixtures().filter((f) =>
    ["silence", "quiet_sine", "reverb_tail", "vocal_like", "dense_music"].includes(f.name),
  );
}

export { SR };
